system/i2c_universal.c

```diff
@@ -0,0 +1,197 @@
+// SPDX-License-Identifier: GPL-2.0
+// Copyright (C) 2024 Loongson Technology Corporation Limited. All rights reserved.
+
+#include "display.h"
+
+#include "io.h"
+#include "string.h"
+
+#include "cpuid.h"
+#include "memrw.h"
+#include "vmem.h"
+#include "smbus.h"
+#include "i2c.h"
+#include "spd.h"
+
+i2c_param i2c_info;
+
+uint8_t max_mc_nu = 0;
+
+//------------------------------------------------------------------------------
+// Private Functions
+//------------------------------------------------------------------------------
+
+static bool determined_i2c_address(void)
+{
+    uint8_t i;
+
+    if (strstr(cpuid_info.vendor_id.str, "Loongson")) {
+        if (strstr(cpuid_info.brand_id.str, "3A5") ||
+              strstr(cpuid_info.brand_id.str, "3A6")) {
+            max_mc_nu = 2;
+            for(i = 0; i < max_mc_nu; i++) {
+                i2c_info.i2c_mc[i].i2c_base = (uint8_t *)map_region(LOONGSON_I2C0_ADDR, 0x8, true);
+                i2c_info.i2c_mc[i].devid.slot0_addr = i * 2 + 0x0;
+                i2c_info.i2c_mc[i].devid.slot1_addr = i * 2 + 0x1;
+           }
+        } else if (strstr((const char *)cpuid_info.brand_id.str, "3C5")) {
+            max_mc_nu = 4;
+            for(i = 0; i < max_mc_nu; i++) {
+                i2c_info.i2c_mc[i].i2c_base = (uint8_t *)map_region(LOONGSON_I2C0_ADDR, 0x8, true);
+            }
+            i2c_info.i2c_mc[0].devid.slot0_addr = 0x4;
+            i2c_info.i2c_mc[0].devid.slot1_addr = 0x5;
+
+            i2c_info.i2c_mc[1].devid.slot0_addr = 0x0;
+            i2c_info.i2c_mc[1].devid.slot1_addr = 0x1;
+
+            i2c_info.i2c_mc[2].devid.slot0_addr = 0x2;
+            i2c_info.i2c_mc[2].devid.slot1_addr = 0x3;
+
+            i2c_info.i2c_mc[3].devid.slot0_addr = 0x6;
+            i2c_info.i2c_mc[3].devid.slot1_addr = 0x7;
+        } else if (strstr(cpuid_info.brand_id.str, "3D5")) {
+            max_mc_nu = 4;
+            i2c_info.i2c_mc[0].i2c_base = (uint8_t *)map_region(LOONGSON_I2C1_ADDR, 0x8, true);
+            i2c_info.i2c_mc[1].i2c_base = (uint8_t *)map_region(LOONGSON_I2C1_ADDR, 0x8, true);
+
+            i2c_info.i2c_mc[2].i2c_base = (uint8_t *)map_region(LOONGSON_I2C2_ADDR, 0x8, true);
+            i2c_info.i2c_mc[3].i2c_base = (uint8_t *)map_region(LOONGSON_I2C2_ADDR, 0x8, true);
+            for(i = 0; i < max_mc_nu; i++) {
+                i2c_info.i2c_mc[i].devid.slot0_addr = (i % 2 * 2 + 0x0);
+                i2c_info.i2c_mc[i].devid.slot1_addr = (i % 2 * 2 + 0x1);
+            }
+        }
+        return true;
+    } else {
+        return false;
+    }
+    return false;
+}
+
+static uint8_t i2c_read_byte(uint8_t *base, uint8_t dev_addr, uint16_t offset)
+{
+    uint8_t buf;
+    volatile uint8_t spd_adr = (uint8_t)(offset & 0xff);
```

Comment:
Because this file was meged into the i2c.c in previous commit, both x86 and LA use the same file, and the CI of the i386 platform cannot pass, because i386 passes the `offset` via a register, the `offset` size mismatch will be reported during build, so I added `volatile` to prevent compiler optimization.